File: app/src/main/java/com/recursia/popularmovies/utils/Mapper.java
package com.recursia.popularmovies.utils;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

abstract public class Mapper<T, V> {

    public final List<V> transform(List<T> list) {
        List<V> collection;
        if (isValid(list)) {
            collection = new ArrayList<>();
            for (T value : list) {
                collection.add(transform(value));
            }
        } else {
            collection = Collections.emptyList();
        }
        return collection;
    }

    abstract public V transform(T t);

    private boolean isValid(Collection collection) {
        return (collection != null) && !(collection.isEmpty());
    }
}

File: app/src/main/java/com/recursia/popularmovies/utils/Mapper.kt
package com.recursia.popularmovies.utils

import java.util.*
import java.util.Collections.emptyList

abstract class Mapper<T, V> {

    fun transform(list: List<T>?): List<V> {
        val collection: MutableList<V>
        if (isValid(list)) {
            collection = ArrayList()
            for (value in list!!) {
                collection.add(transform(value))
            }
        } else {
            collection = emptyList()
        }
        return collection
    }

    private fun isValid(collection: Collection<*>?): Boolean {
        return collection != null && !collection.isEmpty()
    }

    abstract fun transform(t: T): V
}
